File: apps/api/src/Activities/Turbo.Activities.XcSki.Contracts/value/XcSkiConditionsReport.cs
using System.Text.Json.Serialization;
using Turboapi.Activities.value;

namespace Turboapi.Activities.XcSki.value;

public sealed record XcSkiConditionsReport
{
    [JsonPropertyName("activityId")] public Guid ActivityId { get; init; }
    [JsonPropertyName("validAt")] public DateTimeOffset ValidAt { get; init; }
    [JsonPropertyName("fetchedAt")] public DateTimeOffset FetchedAt { get; init; }
    [JsonPropertyName("weather")] public WeatherSlice Weather { get; init; }

    /// <summary>Most-recent grooming status from the external feed
    /// (e.g. Skisporet). Null until the SkisporetGroomingProvider lands;
    /// when null the report falls back to the stored
    /// <c>XcSkiDetails.GroomingStatus</c> field.</summary>
    [JsonPropertyName("liveGroomingHoursAgo")] public int? LiveGroomingHoursAgo { get; init; }

    [JsonPropertyName("score")] public int? Score { get; init; }
    [JsonPropertyName("rationale")] public string Rationale { get; init; }

    [JsonConstructor]
    public XcSkiConditionsReport(
        Guid activityId, DateTimeOffset validAt, DateTimeOffset fetchedAt,
        WeatherSlice weather, int? liveGroomingHoursAgo, int? score, string rationale)
    {
        ActivityId = activityId;
        ValidAt = validAt;
        FetchedAt = fetchedAt;
        Weather = weather;
        LiveGroomingHoursAgo = liveGroomingHoursAgo;
        Score = score;
        Rationale = rationale ?? throw new ArgumentNullException(nameof(rationale));
    }
}

File: apps/flutter/lib/features/activity_xc_ski/models/xc_ski_conditions_report.dart
import 'package:turbo/features/activity_fishing/api.dart' show WeatherSlice;

class XcSkiConditionsReport {
  final String activityId;
  final DateTime validAt;
  final DateTime fetchedAt;
  final WeatherSlice weather;

  /// Hours since the last grooming pass per the live feed (Skisporet).
  /// Null when no live feed is wired up for this trail.
  final int? liveGroomingHoursAgo;

  final int? score;
  final String rationale;

  const XcSkiConditionsReport({
    required this.activityId,
    required this.validAt,
    required this.fetchedAt,
    required this.weather,
    required this.liveGroomingHoursAgo,
    required this.score,
    required this.rationale,
  });

  factory XcSkiConditionsReport.fromJson(Map<String, dynamic> json) =>
      XcSkiConditionsReport(
        activityId: json['activityId'] as String,
        validAt: DateTime.parse(json['validAt'] as String),
        fetchedAt: DateTime.parse(json['fetchedAt'] as String),
        weather: WeatherSlice.fromJson(json['weather'] as Map<String, dynamic>),
        liveGroomingHoursAgo: (json['liveGroomingHoursAgo'] as num?)?.toInt(),
        score: (json['score'] as num?)?.toInt(),
        rationale: json['rationale'] as String,
      );
}
